// opentelemetry-exporter-datadog/build/src/defaults.js
"use strict";
/*
 * Unless explicitly stated otherwise all files in this repository are licensed
 * under the Apache 2.0 license (see LICENSE).
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2020 Datadog, Inc.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatadogInstrumentationTypes = exports.DatadogOtelInstrumentations = exports.DatadogPropagationDefaults = exports.DatadogExportDefaults = exports.DatadogSamplingCodes = exports.DatadogDefaults = void 0;
var DatadogDefaults;
(function (DatadogDefaults) {
    /** The datadog formatted origin tag */
    DatadogDefaults["DD_ORIGIN"] = "_dd_origin";
    /** The otel compliant formatted origin tag */
    DatadogDefaults["OT_ALLOWED_DD_ORIGIN"] = "dd_origin";
    /** The otel tracestate key */
    DatadogDefaults["TRACE_STATE"] = "traceState";
    /** The datadog formatted sample rate key */
    DatadogDefaults["SAMPLE_RATE_METRIC_KEY"] = "_sample_rate";
    /** The otel attribute for error type */
    DatadogDefaults["ERR_NAME_SUBSTRING"] = ".error_name";
    /** The datadog env tag */
    DatadogDefaults["ENV_KEY"] = "env";
    /** The datadog version tag */
    DatadogDefaults["VERSION_KEY"] = "version";
    /** The datadog error tag name */
    DatadogDefaults["ERROR_TAG"] = "error";
    /** A datadog error tag's value */
    DatadogDefaults[DatadogDefaults["ERROR"] = 1] = "ERROR";
    /** The datadog error type tag name */
    DatadogDefaults["ERROR_TYPE_TAG"] = "error.type";
    /** A datadog error type's default value */
    DatadogDefaults["ERROR_TYPE_DEFAULT_VALUE"] = "ERROR";
    /** The datadog error message tag name */
    DatadogDefaults["ERROR_MSG_TAG"] = "error.msg";
    /** The datadog span kind tag name */
    DatadogDefaults["SPAN_KIND"] = "span.kind";
    /** The datadog span type tag name */
    DatadogDefaults["SPAN_TYPE"] = "span.type";
    /** The datadog resource tag name */
    DatadogDefaults["RESOURCE_TAG"] = "resource.name";
    /** The datadog service tag name */
    DatadogDefaults["SERVICE_TAG"] = "service.name";
    /** The datadog span kind client name */
    DatadogDefaults["CLIENT"] = "client";
    /** The datadog span kind server name */
    DatadogDefaults["SERVER"] = "server";
    /** The datadog span kind consumer name */
    DatadogDefaults["PRODUCER"] = "worker";
    /** The datadog span kind consumer name */
    DatadogDefaults["CONSUMER"] = "worker";
    /** The datadog span kind internal name */
    DatadogDefaults["INTERNAL"] = "internal";
    /** The otel traceId name */
    DatadogDefaults["TRACE_ID"] = "traceId";
    /** The otel spanId name */
    DatadogDefaults["SPAN_ID"] = "spanId";
    /** The otel http method attribute name */
    DatadogDefaults["HTTP_METHOD"] = "http.method";
    /** The otel http target attribute name */
    DatadogDefaults["HTTP_TARGET"] = "http.target";
})(DatadogDefaults = exports.DatadogDefaults || (exports.DatadogDefaults = {}));
var DatadogSamplingCodes;
(function (DatadogSamplingCodes) {
    /** The datadog force drop value */
    DatadogSamplingCodes[DatadogSamplingCodes["USER_REJECT"] = -1] = "USER_REJECT";
    /** The datadog auto drop value */
    DatadogSamplingCodes[DatadogSamplingCodes["AUTO_REJECT"] = 0] = "AUTO_REJECT";
    /** The datadog auto keep value */
    DatadogSamplingCodes[DatadogSamplingCodes["AUTO_KEEP"] = 1] = "AUTO_KEEP";
})(DatadogSamplingCodes = exports.DatadogSamplingCodes || (exports.DatadogSamplingCodes = {}));
var DatadogExportDefaults;
(function (DatadogExportDefaults) {
    /** The default service name. */
    DatadogExportDefaults["SERVICE_NAME"] = "dd-service";
    /** The default datadog agent url. */
    DatadogExportDefaults["AGENT_URL"] = "http://localhost:8126";
    /**
     * The default flush interval for the
     * Datadog exporter
     */
    DatadogExportDefaults[DatadogExportDefaults["FLUSH_INTERVAL"] = 1000] = "FLUSH_INTERVAL";
})(DatadogExportDefaults = exports.DatadogExportDefaults || (exports.DatadogExportDefaults = {}));
var DatadogPropagationDefaults;
(function (DatadogPropagationDefaults) {
    /** The default datadog trace id header*/
    DatadogPropagationDefaults["X_DD_TRACE_ID"] = "x-datadog-trace-id";
    /** The default datadog parent span id header*/
    DatadogPropagationDefaults["X_DD_PARENT_ID"] = "x-datadog-parent-id";
    /** The default datadog sampling priority header*/
    DatadogPropagationDefaults["X_DD_SAMPLING_PRIORITY"] = "x-datadog-sampling-priority";
    /** The default datadog origin header*/
    DatadogPropagationDefaults["X_DD_ORIGIN"] = "x-datadog-origin";
})(DatadogPropagationDefaults = exports.DatadogPropagationDefaults || (exports.DatadogPropagationDefaults = {}));
// otel automatic instrumentation library names
var DatadogOtelInstrumentations;
(function (DatadogOtelInstrumentations) {
    DatadogOtelInstrumentations["GRPC"] = "@opentelemetry/plugin-grpc";
    DatadogOtelInstrumentations["HTTP"] = "@opentelemetry/plugin-http";
    DatadogOtelInstrumentations["HTTPS"] = "@opentelemetry/plugin-https";
    DatadogOtelInstrumentations["EXPRESS"] = "@opentelemetry/plugin-express";
    DatadogOtelInstrumentations["IOREDIS"] = "@opentelemetry/plugin-ioredis";
    DatadogOtelInstrumentations["MONGODB"] = "@opentelemetry/plugin-mongodb";
    DatadogOtelInstrumentations["MYSQL"] = "@opentelemetry/plugin-mysql";
    DatadogOtelInstrumentations["PGPOOL"] = "@opentelemetry/plugin-pg-pool";
    DatadogOtelInstrumentations["PG"] = "@opentelemetry/plugin-pg";
    DatadogOtelInstrumentations["REDIS"] = "@opentelemetry/plugin-redis";
})(DatadogOtelInstrumentations = exports.DatadogOtelInstrumentations || (exports.DatadogOtelInstrumentations = {}));
// datadog instrumentation library types
var DatadogInstrumentationTypes;
(function (DatadogInstrumentationTypes) {
    DatadogInstrumentationTypes["WEB"] = "web";
    DatadogInstrumentationTypes["HTTP"] = "http";
    DatadogInstrumentationTypes["MONGODB"] = "mongodb";
    DatadogInstrumentationTypes["REDIS"] = "redis";
    DatadogInstrumentationTypes["SQL"] = "sql";
})(DatadogInstrumentationTypes = exports.DatadogInstrumentationTypes || (exports.DatadogInstrumentationTypes = {}));
