// opentelemetry-exporter-datadog/build/src/types.js
"use strict";
/*
 * Unless explicitly stated otherwise all files in this repository are licensed
 * under the Apache 2.0 license (see LICENSE).
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2020 Datadog, Inc.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.datadog = exports.NoopTracer = exports.Span = exports.Sampler = exports.PrioritySampler = exports.format = exports.AgentExporter = exports.SpanContext = exports.id = void 0;
/* eslint-disable @typescript-eslint/no-var-requires */
exports.id = require('dd-trace/packages/dd-trace/src/id');
exports.SpanContext = require('dd-trace/packages/dd-trace/src/opentracing/span_context');
exports.AgentExporter = require('dd-trace/packages/dd-trace/src/exporters/agent');
exports.format = require('dd-trace/packages/dd-trace/src/format');
exports.PrioritySampler = require('dd-trace/packages/dd-trace/src/priority_sampler');
exports.Sampler = require('dd-trace/packages/dd-trace/src/sampler');
exports.Span = require('dd-trace/packages/dd-trace/src/opentracing/span');
exports.NoopTracer = require('dd-trace/packages/dd-trace/src/noop/tracer');
exports.datadog = require('dd-trace');
/* eslint-enable @typescript-eslint/no-var-requires */
